File: Samples/EnumerateTiffTags/C#/EnumerateTiffTags.cs
using System.Diagnostics;
using System.IO;
using System.Windows.Forms;

using BitMiracle.LibTiff.Classic;

namespace BitMiracle.LibTiff.Samples
{
    public static class EnumerateTiffTags
    {
        public static void Main()
        {
            const string fileName = @"Sample data\multipage.tif";

            using (Tiff image = Tiff.Open(fileName, "r"))
            {
                if (image == null)
                {
                    MessageBox.Show("Could not open incoming image");
                    return;
                }

                using (StreamWriter writer = new StreamWriter("EnumerateTiffTags.txt"))
                {
                    short numberOfDirectories = image.NumberOfDirectories();
                    for (short d = 0; d < numberOfDirectories; ++d)
                    {
                        if (d != 0)
                            writer.WriteLine("---------------------------------");

                        image.SetDirectory((short)d);

                        writer.WriteLine("Image {0}, page {1} has following tags set:\n", fileName, d);
                        for (ushort t = ushort.MinValue; t < ushort.MaxValue; ++t)
                        {
                            TiffTag tag = (TiffTag)t;
                            FieldValue[] value = image.GetField(tag);
                            if (value != null)
                            {
                                for (int j = 0; j < value.Length; j++)
                                {
                                    writer.WriteLine("{0}", tag.ToString());
                                    writer.WriteLine("{0} : {1}", value[j].Value.GetType().ToString(), value[j].ToString());
                                }

                                writer.WriteLine();
                            }
                        }
                    }
                }
            }

            Process.Start("EnumerateTiffTags.txt");
        }
    }
}

File: Samples/EnumerateTiffTags/VB.NET/EnumerateTiffTags.vb
Imports Microsoft.VisualBasic
Imports System.Diagnostics
Imports System.IO
Imports System.Windows.Forms

Imports BitMiracle.LibTiff.Classic

Namespace BitMiracle.LibTiff.Samples
    Public NotInheritable Class EnumerateTiffTags
        Private Sub New()
        End Sub
        Public Shared Sub Main()
            Const fileName As String = "Sample data\multipage.tif"

            Using image As Tiff = Tiff.Open(fileName, "r")
                If image Is Nothing Then
                    MessageBox.Show("Could not open incoming image")
                    Return
                End If

                Using writer As New StreamWriter("EnumerateTiffTags.txt")
                    Dim numberOfDirectories As Short = image.NumberOfDirectories()
                    For d As Short = 0 To numberOfDirectories - 1
                        If d <> 0 Then
                            writer.WriteLine("---------------------------------")
                        End If

                        image.SetDirectory(CShort(d))

                        writer.WriteLine("Image {0}, page {1} has following tags set:" & vbCrLf, fileName, d)
                        For t As Integer = 0 To 65535
                            Dim tag As TiffTag = DirectCast(t, TiffTag)
                            Dim value As FieldValue() = image.GetField(tag)
                            If value IsNot Nothing Then
                                For j As Integer = 0 To value.Length - 1
                                    writer.WriteLine("{0}", tag.ToString())
                                    writer.WriteLine("{0} : {1}", value(j).Value.[GetType]().ToString(), value(j).ToString())
                                Next

                                writer.WriteLine()
                            End If
                        Next
                    Next
                End Using
            End Using

            Process.Start("EnumerateTiffTags.txt")
        End Sub
    End Class
End Namespace